Reduce provider/resource resolution stages down to 1

jkeiser opened the pull request:
- Adds `platform_version` filter to NodeMap, using the Chef::Version::Platform class to parse:
```ruby
class R < Chef::Resource
  provides :r, platform: 'ubuntu', platform_version: '>= 10.1'
end
```
- Adds specificity preferences to NodeMap:
```ruby
class A < Chef::Resource
  provides :a, platform: 'ubuntu', platform_version: '>= 12.0'
end
class B < Chef::Resource
  provides :a, platform: 'ubuntu'
end
class C < Chef::Resource
  provides :a
end
class BOverride < Chef::Resource
  provides :a, platform: 'ubuntu'
end
# No matter what order the above are declared in, preference order is A > B > C
# BOverride is declared *after* B, and so is preferred over B.
```
- Makes Resource/ProviderPriorityMap look through *all* matching arrays for a supported provider, not just the first matching array.  If `provides :service, platform: 'ubuntu'` fails to implement `provides?`, we still search `provides :service, os: 'linux'`.
- Removes node-specific NodeMaps and use the global Resource/ProviderPriorityMaps to implement `provides`
- Removes all ProviderMapping entries and uses `provides` instead
- Moves all requires of providers and resources out of ProviderMapping and Resource/ProviderPriorityMaps and into the relevant classes

thommay commented on lib/chef/provider/mount.rb:
```diff
@@ -24,7 +24,8 @@
   class Provider
     class Mount < Chef::Provider
-
+      Mount.provides :mount
```
why is this sometimes `Foo.provides :foo` and other times `provides :foo`?

jkeiser replied:
Copypasta.  It should never have to do Mount.provides unless you're doing it from outside the class.  I'll search through and fix.

jkeiser replied:
@thommay fixed.

lamont-granquist commented on lib/chef/provider/group/gpasswd.rb:
```diff
@@ -22,6 +22,7 @@ class Chef
   class Provider
     class Group
       class Gpasswd < Chef::Provider::Group::Groupadd
+        provides :group, platform: %w(suse default)
```
There's no `default` platform.   This should probably be `os: linux` which via your herirarchy of specificity would get overridden by more precise `platform` or `platform_version` overrides.  Although this might simply be `provides :group` without any specificity since many OSen would be getting this via the default in the platform_map (everything that doesn't declare a group provider needs this provider).

jkeiser replied:
Thanks! I decided on `provides :group`.

lamont-granquist commented on lib/chef/provider/ifconfig/redhat.rb:
```diff
@@ -22,6 +22,7 @@ class Chef
   class Provider
     class Ifconfig
       class Redhat < Chef::Provider::Ifconfig
+        provides :ifconfig, platform: %w(centos fedora redhat ibm_powerkvm cloudlinux parallels)
```
use platform_family here to tighten this up to fedora and redhat?

jkeiser replied:
i.e. remove that line and add `platform_family: %w(fedora redhat)`?

jkeiser replied:
Are all those platforms variants of fedora and redhat?

lamont-granquist replied:
yes to both.  and you're missing oracle and scientific, both of which should be here, which is why platform_family is better.

lamont-granquist replied:
and it should be 'fedora' and 'rhel', not 'redhat'

lamont-granquist commented on lib/chef/provider/group/usermod.rb:
```diff
@@ -24,6 +24,8 @@ class Group
 
         provides :group, os: "openbsd"
+        provides :group, platform: %w(opensuse), platform_version: '>= 12.3'
+        provides :group, platform: %w(openindiana opensolaris nexentacore omnios solaris2 smartos hpux)
```
it looks like you can use `os: %w{solaris2 hpux}` here

jkeiser replied:
Again, not super sure myself, but happy to make the change :)  (I did this all by carefully keeping everything the same when I moved them.

lamont-granquist replied:
Yeah, the old platform map was buggy, though, because it was based on platform and couldn't do platform_family or os.

lamont-granquist commented on lib/chef/provider/mount/solaris.rb:
```diff
@@ -27,6 +27,8 @@ class Provider
     class Mount
       # Mount Solaris File systems
       class Solaris < Chef::Provider::Mount
+        provides :mount, platform: %w(openindiana opensolaris nexentacore omnios solaris2 solaris2 < 5.11 smartos)
```
"solaris2 < 5.11" doesn't look valid here.

jkeiser replied:
Ooh, thanks.  There were a few like that, thought I had them all.

lamont-granquist commented on lib/chef/provider/package.rb:
```diff
@@ -490,6 +490,38 @@ def allow_downgrade
+
+      Chef.set_provider_priority_array :package, [ Homebrew, Macports ], os: "darwin"
+
+      Chef.set_provider_priority_array :package,  Dpkg,     os: "linux"
```
dpkg isn't a default for all linuxen, this should only be `platform_family: "debian"`

lamont-granquist commented on lib/chef/provider/package.rb:
```diff
@@ -490,6 +490,38 @@ def allow_downgrade
+      Chef.set_provider_priority_array :package, [ Homebrew, Macports ], os: "darwin"
+
+      Chef.set_provider_priority_array :package,  Dpkg,     os: "linux"
+      Chef.set_provider_priority_array :package,  Apt,      platform: %w(ubuntu gcel linaro raspbian linuxmint debian)
```
`platform_family: %w{rhel fedora}`

jkeiser replied:
I'm confused about these last two ... what exactly should be on dpkg and what should be on apt?  It's weird that apt and dpkg would both say debian right now.

lamont-granquist commented on lib/chef/provider/package.rb:
```diff
@@ -490,6 +490,38 @@ def allow_downgrade
+
+      Chef.set_provider_priority_array :package,  Dpkg,     os: "linux"
+      Chef.set_provider_priority_array :package,  Apt,      platform: %w(ubuntu gcel linaro raspbian linuxmint debian)
+      Chef.set_provider_priority_array :package,  Yum,      platform: %w(xenserver xcp centos amazon scientific fedora oracle redhat ibm_powerkvm cloudlinux parallels)
```
`platform_family: %w{rhel fedora}`

lamont-granquist commented on lib/chef/provider/package.rb:
```diff
@@ -490,6 +490,38 @@ def allow_downgrade
+      Chef.set_provider_priority_array :package,  Dpkg,     os: "linux"
+      Chef.set_provider_priority_array :package,  Apt,      platform: %w(ubuntu gcel linaro raspbian linuxmint debian)
+      Chef.set_provider_priority_array :package,  Yum,      platform: %w(xenserver xcp centos amazon scientific fedora oracle redhat ibm_powerkvm cloudlinux parallels)
+      Chef.set_provider_priority_array :package,  Zypper,   platform: %w(opensuse suse)
```
`platform_family: "suse"` i think

lamont-granquist commented on lib/chef/util/path_helper.rb:
```diff
@@ -16,11 +16,208 @@
-
 class Chef
   class Util
-    PathHelper = ChefConfig::PathHelper
```
You've got some kind of rebase conflict here you resolved incorrectly?

jkeiser replied:
Oy!  Restoring.

lamont-granquist commented on lib/chef/platform/provider_mapping.rb:
```diff
@@ -34,267 +29,7 @@ class << self
-              :user => Chef::Provider::User::Useradd,
-              :group => Chef::Provider::Group::Gpasswd,
-              :ifconfig => Chef::Provider::Ifconfig,
-            }
```
You should probably leave the default values here for now.  I don't think we have good enough coverage in the provider resolver unit tests to prove that we won't still ever need to hit this.

jkeiser replied:
Hmm.  I can do that.  Though I'll say, everything that I moved incorrectly, there were tests for.

lamont-granquist commented:
If you revert the cleanup changes that you made to the provider_resolver unit specs does it all still pass?  I think that you didn't modify the logic at all, but its a very large patch.  You should still get green before and after the changes to that file as far as I can see.

jkeiser commented:
@lamont-granquist ran the prior version of the provider_resolver specs, all green.

jkeiser commented:
@lamont-granquist all changes made with the exception of restoring the default provider_mapping, which should be straightforward.

lamont-granquist commented:
:+1:

tyler-ball commented on lib/chef/dsl/recipe.rb:
```diff
@@ -36,58 +37,6 @@ module Recipe
       include Chef::Mixin::PowershellOut
 
-      # method_missing must live for backcompat purposes until Chef 13.
-      def method_missing(method_symbol, *args, &block)
```
:tada:

tyler-ball commented on lib/chef/node_map.rb:
```diff
@@ -68,74 +70,90 @@ def set(key, value, filters = {}, &block)
+    # @return [Object] Value
+    #
+    def list(node, key)
+      # FIXME: real exception
```
It would be nice to see this FIXME solved instead of propogated

tyler-ball commented on spec/unit/provider_resolver_spec.rb:
```diff
@@ -50,6 +54,51 @@
   end
 
+  def self.on_platform(platform, *tags,
+    platform_version: '11.0.1',
```
why is `platform_version` defaulted here?

jkeiser replied:
The others are as well, just they are defaulted to other parameters (see the call later on where we use ||)

jkeiser commented:
@lamont-granquist I have added tests for all the provider mappings and they passed (found a few bugs, see the latest commit).  I went and ran them against master as well, and [here are the changes](https://github.com/chef/chef/commit/3306ee516061ccaed278a5c497f9c51347f970e7) I made to master to get them to pass (indicating that these are changes the provider resolver branch is making now):

- Add Ifconfig::Redhat as providers for oracle, scientific, amazon, xcp and xenserver instead of the base Ifconfig provider.
- Have Package::Paludis provide :package on the exherbo platform.
- Have :portage_package provided by Package::Portage.

I don't actually know whether the first two things are correct or not; the ifconfig thing in particular was one of the things we changed around here with platform_family instead of platform, I think.

- Stop the provider_mapping from modifying its input (it was doing name.downcase! instead of name = name.downcase).  Frozen strings in my constants caught this.
- Make platform_provider

jkeiser commented:
If you can let me know whether the ifconfig and package provider changes are positive or negative, I will get this puppy in.

lamont-granquist commented:
Those changes look good.  The redhat stuff looks like a bugfix falling out of this cleanup rather than a backcompat break.

:shipit:
